hnn_core/basket.py

```diff
@@ -89,45 +64,49 @@ def parreceive(self, gid, gid_dict, pos_dict, p_ext):
         for gid_src, p_src, pos in zip(gid_dict['common'],
                                        p_ext, pos_dict['common']):
             # check if AMPA params are defined in the p_src
-            if 'L2Basket_ampa' in p_src.keys():
+            if f'{self.name}_ampa' in p_src.keys():
                 # create an nc_dict
-                nc_dict_ampa = {
+                nc_dict = {
                     'pos_src': pos,
-                    'A_weight': p_src['L2Basket_ampa'][0],
-                    'A_delay': p_src['L2Basket_ampa'][1],
+                    'A_weight': p_src[f'{self.name}_ampa'][0],
+                    'A_delay': p_src[f'{self.name}_ampa'][1],
                     'lamtha': p_src['lamtha'],
                     'threshold': p_src['threshold'],
                     'type_src': 'ext'
                 }
 
-                # AMPA synapse
-                self.ncfrom_common.append(self.parconnect_from_src(
-                    gid_src, nc_dict_ampa, self.soma_ampa))
+                self._connect_feed_at_loc(
+                    feed_loc='proximal', receptor='ampa',
+                    gid_src=gid_src, nc_dict=nc_dict,
+                    nc_list=self.ncfrom_common)
 
             # Check if NMDA params are defined in p_src
-            if 'L2Basket_nmda' in p_src.keys():
-                nc_dict_nmda = {
+            if f'{self.name}_nmda' in p_src.keys():
+                nc_dict = {
                     'pos_src': pos,
-                    'A_weight': p_src['L2Basket_nmda'][0],
-                    'A_delay': p_src['L2Basket_nmda'][1],
+                    'A_weight': p_src[f'{self.name}_nmda'][0],
+                    'A_delay': p_src[f'{self.name}_nmda'][1],
                     'lamtha': p_src['lamtha'],
                     'threshold': p_src['threshold'],
                     'type_src': 'ext'
                 }
 
-                # NMDA synapse
-                self.ncfrom_common.append(self.parconnect_from_src(
-                    gid_src, nc_dict_nmda, self.soma_nmda))
+                self._connect_feed_at_loc(
+                    feed_loc='proximal', receptor='nmda',
```

Comment:
Why does `feed_loc='proximal'` in all of the `BasketSingle._connect_feed_at_loc()` calls? I feel like I'm missing something really obvious...

Reply:
This is a very good point. So looking at the [documentation of HNN](https://jonescompneurolab.github.io/hnn-under_the_hood/04_evoked-rhythmic-inputs/04_evoked-rhythmic-inputs), it says this for proximal:

> The left schematic here shows proximal inputs which target basal dendrites of layer 2/3 and layer 5 pyramidal neurons, and somata of layer 2/3 and layer 5 interneurons.

and this for distal:

> The right schematic shows distal inputs which target the distal apical dendrites of layer 5 and layer 2/3 pyramidal neurons and the somata of layer 2/3 interneurons

*So, for layer 2/3 there is no difference between proximal and distal. For layer 5, you have an input drive to somata for proximal but not distal*

Now if you go the file [L5_basket.py](https://github.com/jonescompneurolab/hnn/blob/master/L5_basket.py) (I'm using the HNN repository since that's our "gold standard" implementation) and look for `'loc'` there is no reference to it. If you go to [L2_basket.py](https://github.com/jonescompneurolab/hnn/blob/master/L2_basket.py#L159-L167), there is an if-else clause but both branches have **exactly** the same lines (right??) with a cryptic comment which asks "why only L2 basket and not L5 basket". Am I missing something?

Reply:
I agree with everything you said @jasmainak. It appears that there is no difference between distal and proximal in the code. It looks like the code changed from only having combined excitatory inputs to both ampa/nmda in the following two commits (in which the comment was added). Commit [317d3324](https://github.com/jonescompneurolab/hnn/commit/317d332407071f1dd5baba66c6bf00ef80c08075#diff-adc0f6cc5cc1a0937da7a9202d463707R159-R164) explains why there was initially an if/else clause. Then the branches became the same in commit [3eee0bd3](https://github.com/jonescompneurolab/hnn/commit/3eee0bd35015a5bbd2bd7761a1c3f1fce8cedd12#diff-adc0f6cc5cc1a0937da7a9202d463707R160-R167).

I think the comment is just reflecting what you observed @jasmainak.

Additionally, a similar comment in [L5_basket.py](https://github.com/jonescompneurolab/hnn/commit/3eee0bd35015a5bbd2bd7761a1c3f1fce8cedd12#diff-0c776e7033b7f5dc6378a67be72276f7R147) noted the difference, but the code still the same thing for proximal and distal inputs. The desired effect comes from there not being a 'L5_basket' key in `p_ext` for distal evoked inputs (in [paramrw.py](https://github.com/jonescompneurolab/hnn/blob/b1522b9e517bc110131ecd2281c9025240980737/paramrw.py#L678-L711)).

Reply:
Yes, we should expect L2/3 basket cells to receive both proximal and distal inputs and L5 basket cells to receive only proximal inputs.

>If you go to L2_basket.py, there is an if-else clause but both branches have exactly the same lines (right??)

Correct, the if/else statement is unnecessary. It looks like this was corrected in [L5_basket.py](https://github.com/jonescompneurolab/hnn/blob/b1522b9e517bc110131ecd2281c9025240980737/L5_basket.py#L146-L172).

>with a cryptic comment which asks "why only L2 basket and not L5 basket". Am I missing something?

This comment is resolved by noting that HNN code *does*  control for proximal vs. distal inputs for both the L2 and L5 basket cell cases. Both [L5_basket](https://github.com/jonescompneurolab/hnn/blob/b1522b9e517bc110131ecd2281c9025240980737/L5_basket.py#L148) and [L2_basket](https://github.com/jonescompneurolab/hnn/blob/b1522b9e517bc110131ecd2281c9025240980737/L2_basket.py#L138) have an 'if' statement that filters feed location-specific connections according to the parameter keys passed via [`p_ext`](https://github.com/jonescompneurolab/hnn/blob/b1522b9e517bc110131ecd2281c9025240980737/paramrw.py#L678-L711). Particularly note that `p_unique['evdist*']` items don't contain a dict key for `L5_basket`.

Reply:
Sorry @blakecaldwell your previous post just loaded for me.

Reply:
wow! this sounds like a Sherlock Holmes detective story ... thanks for the sleuthing both of you.

Let me see what I can do to improve the readability.